### PR TITLE
feat: integrate withdraw SPL

### PR DESCRIPTION
**Pull request by skosito**
# Description

- integrate withdraw spl
- e2e test to withdraw spl to receiver that has associated token account
- e2e test to withdraw spl to receiver that doesn't have ata, testing that gateway's rent payer will create it

missing:
- unit tests, i will revisit unit tests for whole solana functionality and add missing tests in follow up PR, this should be last PR that adds new feature

# How Has This Been Tested?

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions

## Release Notes

- **New Features**
	- Introduced Whitelist message capability for whitelisting SPL tokens on Solana.
	- Added SPL withdrawal and deposit functionalities.
	- Enhanced script for airdropping SOL tokens to support multiple accounts.
	- New end-to-end tests for SPL withdrawals and associated token account creation.

- **Improvements**
	- Enhanced build reproducibility with stable checksums.
	- Expanded testing framework for improved error handling and coverage.

- **Bug Fixes**
	- Resolved issues with inbound connections and omnichain calls.

- **Documentation**
	- Updated changelog to reflect new features, improvements, and fixes.

### CODE REVIEW

File: zetaclient/chains/solana/signer/signer.go
Comment (swift1337) [inline zetaclient/chains/solana/signer/signer.go:343]: Let's remove dev dependency

File: pkg/contracts/solana/gateway_message.go
Comment (swift1337) [inline pkg/contracts/solana/gateway_message.go:132]: What does "Ata" mean?

File: pkg/contracts/solana/gateway_message.go
Comment (swift1337) [inline pkg/contracts/solana/gateway_message.go:194]: Consider using `binary.BigEndian.AppendUint64`, no need for intermediate `buff`

File: e2e/e2etests/test_spl_withdraw.go
Comment (fbac) [inline e2e/e2etests/test_spl_withdraw.go:65]: Strange formatting here, is this the result of `make fmt`?

File: e2e/e2etests/test_spl_withdraw.go
Comment (fbac) [inline e2e/e2etests/test_spl_withdraw.go:33]:
```suggestion
	require.Equal(
		r,
		-1,
		withdrawAmount.Cmp(approvedAmount),
		"Withdrawal amount must be less than the approved amount: %v",
		approvedAmount,
	)
```

File: e2e/e2etests/test_spl_withdraw.go
Comment (fbac) [inline e2e/e2etests/test_spl_withdraw.go:40]: nit: Not really an issue but since this function is related only to Solana, could it be renamed somehow, like `SolanaFindOrCreateATA`. In the end something that shows it's Solana related, to group functions by chain and makes it easier in IDEs to filter for chain functions.

File: e2e/e2etests/test_spl_withdraw_and_create_receiver_ata.go
Comment (fbac) [inline e2e/e2etests/test_spl_withdraw_and_create_receiver_ata.go:35]:
```suggestion
	require.Equal(
		r,
		-1,
		withdrawAmount.Cmp(approvedAmount),
		"Withdrawal amount must be less than the approved amount: %v",
		approvedAmount,
	)
```

File: e2e/runner/setup_solana.go
Comment (fbac) [inline e2e/runner/setup_solana.go:90]: nit: Same as before, since the E2ERunner is multichain it would make sense to group helper functions by name.

File: zetaclient/chains/solana/signer/signer.go
Comment (fbac) [inline zetaclient/chains/solana/signer/signer.go:249]: We could use some defensive approach and check cctx is nil before using it. Same for params as it's a pointer.

File: zetaclient/chains/solana/signer/signer.go
Comment (fbac) [reply zetaclient/chains/solana/signer/signer.go:331]: In that case maybe leave a comment with the url referencing the function, so if it changes in the future we could check it easily?

File: zetaclient/chains/solana/signer/withdraw_spl.go
Comment (fbac) [inline zetaclient/chains/solana/signer/withdraw_spl.go:31]:
```suggestion
	// zero out the amount if cancelTx is set. It's legal to withdraw 0 spl through the gateway.
```

**Review by fbac — APPROVED** [review]
LGTM, left minor comments.

**Review by lumtis — APPROVED** [review]
Looks good to me. Would be great getting a re-review from @swift1337 before considering merging

File: e2e/runner/solana.go
Comment (swift1337) [inline e2e/runner/solana.go:163]:
```suggestion
// ResolveSolanaATA finds or creates SOL associated token account 
func (r *E2ERunner) ResolveSolanaATA(
```

File: e2e/runner/solana.go
Comment (swift1337) [inline e2e/runner/solana.go:34]:
```suggestion
func (r *E2ERunner) SolanaRentPayerPDA() solana.PublicKey {
```

File: e2e/runner/solana.go
Comment (swift1337) [inline e2e/runner/solana.go:33]:
```suggestion
// ComputePdaAddress computes the rent payer PDA (Program Derived Address) for the gateway program
```

File: zetaclient/chains/solana/signer/withdraw_spl.go
Comment (swift1337) [inline zetaclient/chains/solana/signer/withdraw_spl.go:45]: Consider wrapping errors

File: zetaclient/chains/solana/signer/withdraw_spl.go
Comment (swift1337) [inline zetaclient/chains/solana/signer/withdraw_spl.go:64]: Remove this log line

File: zetaclient/chains/solana/signer/withdraw_spl.go
Comment (swift1337) [inline zetaclient/chains/solana/signer/withdraw_spl.go:162]: consider preallocating the slice:

```go
meta := []*solana.AccountMeta{
    a,
    b,
    c, 
    ....
}

inst.AccountValues = meta
```

File: zetaclient/chains/solana/signer/withdraw_spl.go
Comment (swift1337) [inline zetaclient/chains/solana/signer/withdraw_spl.go:77]: I wanted to suggest that we avoid using this pattern. It makes the code harder to understand and can lead to errors. It's also more complex and spreads the context across multiple functions.

Here's an example of what I mean:

```go
var foo SomeStruct

// ...
foo.a = 123

// ...

foo.b = "hello"

actMyWriteSomeSideEffects(&foo)

// ...

return foo, nil
```

Instead I propose the following:
```go
a := 123

// ...
b := "hello"

// ...

c,d := calculateSomething(...)

// ...

return SomeStruct{
  aa: a,
  bb: b,
  cc: c,
  dd: d,
}, nil
```

File: pkg/contracts/solana/gateway.go
Comment (swift1337) [inline pkg/contracts/solana/gateway.go:64]: IMO this makes code more consistent

```suggestion
func RentPayerPDA(gateway solana.PublicKey) (solana.PublicKey, error) {
```

File: pkg/contracts/solana/gateway.go
Comment (swift1337) [inline pkg/contracts/solana/gateway.go:47]:
```suggestion
func ParseGatewayWithPDA(gatewayAddress string) (solana.PublicKey, solana.PublicKey, error) {
```

File: e2e/e2etests/test_spl_withdraw.go
Comment (ws4charlie) [inline e2e/e2etests/test_spl_withdraw.go:43]: Is it possible to use `rpc.CommitmentFinalized` in E2Es?

File: e2e/e2etests/test_spl_withdraw.go
Comment (ws4charlie) [inline e2e/e2etests/test_spl_withdraw.go:60]: same question

**Review by swift1337 — APPROVED** [review]

**Pull request merged**
